### PR TITLE
don't update parent event changed|failed in bulk (it's expensive for large tables)

### PR DESCRIPTION
**Pull request by ryanpetrello**

### CODE REVIEW

File: awx/main/models/events.py [hunk 1]
```diff
@@ -352,9 +352,19 @@ def save(self, *args, **kwargs):
...
+                self.event.startswith('runner_on'),
+                self.event in self.FAILED_EVENTS
+            ]) and self.parent_uuid:
+                kwargs = {}
+                if self.changed is True:
+                    kwargs['changed'] = True
+                if self.failed is True:
+                    kwargs['failed'] = True
+                if kwargs:
+                    JobEvent.objects.filter(job=self.job, uuid=self.parent_uuid).update(**kwargs)
```
Comment (ryanpetrello) [inline awx/main/models/events.py]:
```sql
\d main_jobevent

...
Indexes:
    ...
    "main_jobevent_job_id_uuid_3df694c5_idx" btree (job_id, uuid)
```

File: awx/main/models/events.py [hunk 2]
```diff
@@ -352,9 +352,19 @@ def save(self, *args, **kwargs):
...
+                self.event.startswith('runner_on'),
+                self.event in self.FAILED_EVENTS
+            ]) and self.parent_uuid:
+                kwargs = {}
+                if self.changed is True:
+                    kwargs['changed'] = True
+                if self.failed is True:
+                    kwargs['failed'] = True
+                if kwargs:
+                    JobEvent.objects.filter(job_id=self.job_id, uuid=self.parent_uuid).update(**kwargs)
```
Comment (ryanpetrello) [inline awx/main/models/events.py:362]:
```sql
\d main_jobevent

...
Indexes:
    ...
    "main_jobevent_job_id_uuid_3df694c5_idx" btree (job_id, uuid)
```

Comment (ryanpetrello) [reply awx/main/models/events.py:362]: Sample query observed running this through a debugger:

```sql
UPDATE "main_jobevent"
    SET "failed" = true
    WHERE ("main_jobevent"."job_id" = 7
    AND "main_jobevent"."uuid" = '0242ac12-0005-4e0c-d27c-00000000000d');
```

Comment (ryanpetrello) [reply awx/main/models/events.py:362]: Sample index scan on a production install:

```sql
awx=> EXPLAIN UPDATE "main_jobevent" SET "failed" = true WHERE ("main_jobevent"."job_id" = 7 AND "main_jobevent"."uuid" = '0242ac12-0005-4e0c-d27c-00000000000d');
                                                      QUERY PLAN
-----------------------------------------------------------------------------------------------------------------------
 Update on main_jobevent  (cost=0.41..8.43 rows=1 width=964)
   ->  Index Scan using main_jobevent_job_id_1a69b01a6e969026_idx on main_jobevent  (cost=0.41..8.43 rows=1 width=964)
         Index Cond: ((job_id = 7) AND ((uuid)::text = '0242ac12-0005-4e0c-d27c-00000000000d'::text))
(3 rows)

awx=> SELECT COUNT(id) FROM main_jobevent;
 count
-------
 34765
(1 row)
```

File: awx/main/models/events.py [hunk 3]
```diff
@@ -352,9 +352,19 @@ def save(self, *args, **kwargs):
         if hasattr(self, 'job') and not from_parent_update:
             if getattr(settings, 'CAPTURE_JOB_EVENT_HOSTS', False):
                 self._update_hosts()
-            if self.event == 'playbook_on_stats':
-                self._update_parents_failed_and_changed()
+            if any([
```
Comment (matburt) [inline awx/main/models/events.py]: I think we can shortcut this entirely by just checking `self.changed` and `self.failed` for `True` and then falling into the JobEvent filter.

We'll save potentially `dozens` of microseconds!

**Review by matburt — CHANGES_REQUESTED** [review]

File: awx/main/models/events.py (hunk 3 above)
Comment (matburt) [reply awx/main/models/events.py]: Even if this means affecting a parent above the task level I don't think that's such a big deal.

**Comment by matburt** [comment]
One other thing will be checking to see if we bubble changed/failed up to a higher level beyond the task level in the current implementation.

**Comment by ryanpetrello** [comment]
> One other thing will be checking to see if we bubble changed/failed up to a higher level beyond the task level in the current implementation.

I wouldn't _expect_ it to, given that we're currently filtering on these events on `.startwith('runner_on')`:

https://github.com/ansible/awx/blob/fc94b3a943d20c82216898e58c979027d80343a3/awx/main/models/events.py#L404

It seems to me like removing the `self.event` string check would result in a behavior change.

**Comment by ryanpetrello** [comment]
Old query is _definitely_ slow super slow for large table sizes:

```sql
awx-dev=> INSERT INTO main_jobevent
    (created, modified, uuid, parent_uuid, failed, changed, host_name, play, role, task, counter, event, job_id, event_data, start_line, end_line, stdout, verbosity, playbook)
    SELECT now(), now(), x, 'parent_uuid', false, false, '', '', '', '', 1, 'playbook_on_start', 1, '{"pid": 3, "playbook_uuid": "461fdef8-ed03-402a-9f90-d283961ec8e3", "playbook": "hello_world.yml"}', 0, 1, 'foo', 0, 'hello_world.yml'
    FROM generate_series(1, 100000000) AS x;

awx-dev=> SELECT COUNT(id) FROM main_jobevent;
-[ RECORD 1 ]---
count | 11000043
```

`<run a simple Job Template that emits events>`

```sql
\timing
UPDATE "main_jobevent"
    SET "failed" = true
    WHERE "main_jobevent"."uuid" IN (
        SELECT U0."parent_uuid" AS Col1
        FROM "main_jobevent" U0
        WHERE (U0."job_id" = 16 AND U0."event"::text LIKE 'runner\_on%' AND U0."failed" = true)
    );
UPDATE 1
Time: 5060.824 ms
```

**Comment by ryanpetrello** [comment]
If you pump the event count up to 25-30M, it can get pretty bad:

```sql
UPDATE "main_jobevent"
    SET "failed" = true
    WHERE "main_jobevent"."uuid" IN (
        SELECT U0."parent_uuid" AS Col1
        FROM "main_jobevent" U0
        WHERE (U0."job_id" = 16 AND U0."event"::text LIKE 'runner\_on%' AND U0."failed" = true)
    );
UPDATE 1
Time: 31314.081 ms
```

**Review by matburt — APPROVED** [review]

**Pull request merged**